lib/portage/sync/modules/git/git.py

```diff
@@ -228,7 +228,39 @@ def update(self):
         if not self.verify_head(revision="refs/remotes/%s" % remote_branch):
             return (1, False)
 
-        if shallow:
+        # Clean up the repo before trying to sync to upstream's
+        clean_cmd = [self.bin_command, "clean", "--force", "-d", "-x"]
+
+        if quiet:
+            clean_cmd.append("--quiet")
+
+        portage.process.spawn(
+            clean_cmd,
+            cwd=portage._unicode_encode(self.repo.location),
+            **self.spawn_kwargs,
+        )
+
+        if exitcode != os.EX_OK:
+            msg = "!!! git clean error in %s" % self.repo.location
+            self.logger(self.xterm_titles, msg)
+            writemsg_level(msg + "\n", level=logging.ERROR, noiselevel=-1)
+            return (exitcode, False)
+
+        # `git diff --quiet` returns 0 on a clean tree and 1 otherwise
+        is_clean = (
+            portage.process.spawn(
+                f"{self.bin_command} diff --quiet",
+                cwd=portage._unicode_encode(self.repo.location),
+                **self.spawn_kwargs,
+            )
+            == 0
+        )
+
+        if not is_clean:
+            # If the repo isn't clean, clobber any changes for parity
+            # with rsync
+            merge_cmd = [self.bin_command, "reset", "--hard"]
```

Comment:
Wouldn't it be better to refrain from using `git reset --hard` and do a `git restore --staged --worktree .` instead and do a `git reset --merge` irrespective of shallowness?

Reply:
git restore's man page says "THIS COMMAND IS EXPERIMENTAL. THE BEHAVIOR MAY CHANGE."

Reply:
k, I didn't notice that part. Ignore my comment on `git reset --merge` for obvious reasons.

Reply:
Wouldn't the combo of `git checkout` and `git clean -fdx` suffice? This would take care of:
- ❌ Deletes local, non-pushed commits
- ✅ Reverts changes you made to tracked files (via `git checkout .`)
- ❌ Restores tracked files you deleted
- ✅ Deletes files/dirs listed in .gitignore (like build files) (via `git clean -fdx`)
- ✅ Deletes files/dirs that are not tracked and not in .gitignore (via `git clean -fdx`)

source: https://stackoverflow.com/a/42903805

Reply:
In this case, it's sufficient to only do a `git reset --hard` after a prior shallow clone/fetch.

Reply:
No, that won't fix things like a diverged tree that we've discussed ad nauseum at this point.

Reply:
k. fyi, I didn't participate in that discussion.